godot-core/src/obj/script.rs

```diff
@@ -417,298 +457,255 @@ impl<'a, T: ScriptInstance> DerefMut for SiMut<'a, T> {
-    /// This pointer must have been created by [super::create_script_instance] and transfered to Godot.
-    unsafe fn instance_data_as_script_instance<'a, T: ScriptInstance>(
-        p_instance: sys::GDExtensionScriptInstanceDataPtr,
-    ) -> &'a ScriptInstanceData<T> {
-        &*(p_instance as *mut ScriptInstanceData<T>)
-    }
-
-    fn transfer_bool_to_godot(value: bool) -> sys::GDExtensionBool {
-        value as sys::GDExtensionBool
-    }
-
-    /// # Safety
-    ///
-    /// - We expect the engine to provide a valid variant pointer the return value can be moved into.
-    unsafe fn transfer_variant_to_godot(variant: Variant, return_ptr: sys::GDExtensionVariantPtr) {
-        variant.move_into_var_ptr(return_ptr)
+    /// Helper struct to store the lengths of lists so they can be properly freed.
+    pub struct PtrlistContainer<T> {
+        list_lengths: Mutex<HashMap<*mut T, u32>>,
     }
 
-    /// # Safety
-    ///
-    /// - The returned `*const T` is guaranteed to point to a list that has an equal length and capacity.
-    fn transfer_ptr_list_to_godot<T>(ptr_list: Box<[T]>, list_length: &mut u32) -> *mut T {
-        *list_length = ptr_list.len() as u32;
-
-        let ptr = Box::into_raw(ptr_list);
-
-        // SAFETY: `ptr` was just created in the line above and should be safe to dereference.
-        unsafe { (*ptr).as_mut_ptr() }
-    }
+    impl<T> PtrlistContainer<T> {
+        pub fn new() -> Self {
+            Self {
+                list_lengths: Mutex::new(HashMap::new()),
+            }
+        }
 
-    /// The returned pointer's lifetime is equal to the lifetime of `script`
-    fn transfer_script_to_godot(script: &Gd<crate::classes::Script>) -> sys::GDExtensionObjectPtr {
-        script.obj_sys()
-    }
+        /// Convert a list into a pointer + length pair. Should be used together with [`list_from_sys`](Self::list_from_sys).
+        ///
+        /// If `list_from_sys` is not called on this list then that will cause a memory leak.
+        pub fn list_into_sys(&self, list: Vec<T>) -> (*const T, u32) {
+            let len: u32 = list
+                .len()
+                .try_into()
+                .expect("list must have length that fits in u32");
+            let ptr = Box::leak(list.into_boxed_slice()).as_mut_ptr();
+
+            let old_value = self.list_lengths.lock().unwrap().insert(ptr, len);
+            assert_eq!(
+                old_value, None,
+                "attempted to insert the same list twice, this is a bug"
+            );
 
-    /// # Safety
-    ///
-    /// - `ptr` is expected to point to a list with both a length and capacity of `list_length`.
-    /// - The list pointer `ptr` must have been created with [`transfer_ptr_list_to_godot`].
-    unsafe fn transfer_ptr_list_from_godot<T>(ptr: *const T, list_length: usize) -> Vec<T> {
-        Vec::from_raw_parts(sys::force_mut_ptr(ptr), list_length, list_length)
-    }
+            (ptr as *const T, len)
```

Comment:
`sys::to_const_ptr`...

or can you call `as_ptr()` instead of `as_mut_ptr()` directly? Does the hashmap need mut pointers?

Reply:
we do need to call `as_mut_ptr()` because otherwise we get a pointer derived from a `&[T]` reference which we cannot later free again since it doesn't have write permissions.

Reply:
I see, thanks for elaborating! Then you could use `sys::to_const_ptr(ptr)` in the return tuple, to avoid `as`.